dartagnan/src/main/java/com/dat3m/dartagnan/analysis/graphRefinement/graphs/eventGraph/axiom/Constraint.java

```diff
@@ -0,0 +1,34 @@
+package com.dat3m.dartagnan.analysis.graphRefinement.graphs.eventGraph.axiom;
+
+import com.dat3m.dartagnan.analysis.graphRefinement.graphs.eventGraph.EventGraph;
+import com.dat3m.dartagnan.analysis.graphRefinement.graphs.eventGraph.GraphListener;
+import com.dat3m.dartagnan.utils.dependable.Dependent;
+import com.dat3m.dartagnan.verification.model.Edge;
+import com.dat3m.dartagnan.verification.model.ExecutionModel;
+
+import java.util.Collections;
+import java.util.List;
+
+public abstract class Constraint implements GraphListener, Dependent<EventGraph> {
+    protected ExecutionModel context;
+    protected final EventGraph constrainedGraph;
+
+    public EventGraph getConstrainedGraph() { return constrainedGraph; }
+
+    public List<EventGraph> getDependencies() {
+        return Collections.singletonList(constrainedGraph);
+    }
+
+    public Constraint(EventGraph constrainedGraph) {
+        this.constrainedGraph = constrainedGraph;
+    }
+
+    public void initialize(ExecutionModel context) {
+        this.context = context;
+    }
+
+    public abstract boolean checkForViolations();
+
+    public abstract List<List<Edge>> getViolations();
```

Comment:
On one hand, the default implementation would need data that is not needed for the ```AcyclicityConstraint``` and on the other hand, I wouldn't want the coupling between ```EmptinessConstraint``` and ```IrreflexivityConstraint``` just because they happen to have similar implementations for two of their methods. Especially cause the implementation is a one-liner so code sharing introduces far more coupling than it actually reduces code. Furthermore, this way you can see the whole implementation in each ```Constraint``` class without checking the base class.